pandas/core/indexes/datetimelike.py

```diff
@@ -279,19 +279,26 @@ def map(self, mapper, na_action=None):
         except Exception:
             return self.astype(object).map(mapper)
 
-    def sort_values(self, return_indexer=False, ascending=True):
+    def sort_values(
+        self, return_indexer=False, ascending=True, key: Optional[Callable] = None
+    ):
         """
         Return sorted copy of Index.
         """
+        if key:
+            idx = self.map(key, na_action="ignore")
+        else:
+            idx = self
+
         if return_indexer:
-            _as = self.argsort()
+            _as = idx.argsort()
             if not ascending:
                 _as = _as[::-1]
             sorted_index = self.take(_as)
             return sorted_index, _as
         else:
-            sorted_values = np.sort(self._ndarray_values)
-            attribs = self._get_attributes_dict()
+            sorted_values = np.sort(idx._ndarray_values)
+            attribs = self._get_attributes_dict()  # type: ignore
```

Comment:
Same comment here on ignore

Reply:
Problem here is `pandas/core/indexes/datetimelike.py:301: error: "DatetimeIndexOpsMixin" has no attribute "_get_attributes_dict"
pandas/core/indexes/datetimelike.py:314: error: "DatetimeIndexOpsMixin" has no attribute "_simple_new"`. Adding the `key` argument causes this, even if no other changes are made. Maybe mypy is getting confused by some inheritance thing?

Reply:
It can be fixed either by adding something like

```
if not isinstance(self, Index):
            raise TypeError("sort_values must be called on an Index object")
```

or by adding methods like `_get_attributes_dict` to the `DatetimeIndexOpsMixin` class.